### ISSUE TITLE
Table CSV download not correct 

### ISSUE THREAD

**Issue by ivokwee** [bigomics/omicsplayground#394]
Many tables do not download correctly. See for example geneset enrichment table below. We need to check every table. This is due to wrong plot_data() passed to csvFunc. Remember, plot_data() should return a clean data.frame/matrix or a list where the *first* element of the list must be a data.frame or matrix. The first element of a list is passed to the downloadHandler. Let's check and list all tables that need a fix.

![image](https://github.com/bigomics/omicsplayground/assets/19613146/9bd3355d-a2c6-4307-af6d-68f15f513906)

**Comment by phisanti** [comment]
I noticed the same problem using the example dataset, but when I commented out the argument `csvFunc = table_data`, it seemed to return the appropriate table. See file `enrichment_table_enrichment_analysis.R`.

```r
    gseatable <- TableModuleServer(
      "datasets",
      func = gseatable.RENDER,
      func2 = gseatable.RENDER_modal,
      #csvFunc = table_data, # This line has to be removed to obtain the appropiate data
      selector = "single"
    )
```

This way, the module `TableModuleServer` (see file ui-TableModule2.R) seems to take the data from `func = gseatable.RENDER` and produce the correct table. However, looking closer at the problem, I realised the `TableModuleServer` has the following logic:

```r
    gseatable <- TableModuleServer(
      "datasets",
      func = gseatable.RENDER,
      func2 = gseatable.RENDER_modal,
      #csvFunc = table_data, # This line has to be removed to obtain the appropriate data
      selector = "single"
    )
```r
      # Downloader
      output$download <- shiny::downloadHandler(
        filename = filename,
        content = function(file) {
          if(!is.null(csvFunc)) {
            dt <- csvFunc()
            print(head(dt))
            if(is.list(dt)) dt <- dt[[1]]
          } else {
            dt <- func()$x$data
          }
          dt2 <- format(dt, digits=4)  ## round
          write.csv(dt2, file=file, row.names=FALSE)
        }
      )

Here, when the function pass a csvFunc and take the dt, it also selects the first column as a vector if it is a list:
```r
            if(is.list(dt)) dt <- dt[[1]]
```
Taking on the account that all `data.frames` are list, then, this line likely generates the bug:
```r
dt <- data.frame(mtcars)
is.list(dt)
# > TRUE
```
If it gets commented out, the downloaded table has the correct content. If there are other lists where it is necessary to extract the first element from the list, it could be possible to solve the issue by using `!is.data.frame(dt)`.

**Comment by ivokwee** [comment]
Good catch. I have included the fix in commit a582d87. The Geneset table now outputs correctly:

![image](https://github.com/bigomics/omicsplayground/assets/19613146/3924c516-0920-497c-b483-7355fa96bc21)
